Get distance between two stations fre 547

brandesdavid opened the pull request:
## Describe the issue:

We need for the new backend a distance api, to get the station count distance between two stations.

## Explain how you solved the issue:

- updated the drizzle snapshot, because of renaming of stations_lines -> lines_stations
-> just db-XXX commands work again

- registered route from getDistance api

- added pathfinding.ts in transit module
-> includes an a* algorithm that takes heuristics of geographical distance in lang and longitude
-> MinHeap to get the closest neighbouring station in each a* iteration

- added transit graph service.ts 
-> includes graph service for loading and getting a* graph, and the function to getDistance

- added transit path cache service.ts to cache already processed distances
-> added precompute of all paths to get an instant O(1) distance find
-> this also saves the computed paths to disk (meaning disk persistence)
-> incase the server is precomputing the paths, we can also just normally compute the distance


## Additional notes or considerations:

~future error handling should put out detailed error outputs. currently, in all apis, we only receive internal server error~
error handling has been integrated

brandesdavid committed:
* feat: add placeholder api request for getDistance
* chore: migrate sql database

brandesdavid committed:
feat:
- add a* path finding algorithm
- add graph cache for faster distance search
- integrate path finding alg to getDistance service

brandesdavid committed:
refactor:
- remove the on demand caching of distance api
- add transit path cache service
feat:
- add precomputing of all nodes to get O(1) requests

brandesdavid committed:
refactor:
- move graph functions from tansit network data service to transit graph service
- added precompute promise

brandesdavid committed:
feat: implement error handling from error handler

johan-t reviewed (changes requested):
Can you fix lint and add some city agnostic integration tests? After that I can start reviewing. Thanks 🙏

johan-t commented on packages/hono-backend/README.md:
```diff
@@ -42,6 +42,22 @@ just db-seed
 
+## Transit path cache (.cache)
+
+The transit distance endpoint stores precomputed path distances in `.cache/transit-paths-cache.json`.
```
In the db would make it simpler, since we do not have to add differences between the file systems, I would argue

johan-t committed:
Merge branch 'main' into get-distance-between-two-stations-fre-547

johan-t committed:
refactor: clean up

johan-t committed:
docs: define response schema

johan-t committed:
feat: test implementation

johan-t committed:
refactor: simplify by using dynamic in memory cache

johan-t commented:
Quick summary of the last four commits and the reasoning behind them:

- Consolidated the distance logic by removing the separate graph service and moving graph loading into `TransitNetworkDataService`. The graph was only used through the distance path, so keeping another service layer added indirection without much benefit.

- Tightened error handling by introducing typed domain errors for pathfinding failures and documenting when to use `AppError`. Lower-level pathfinding now throws typed domain errors, while the service layer translates them into stable API responses like `STATION_NOT_FOUND` and `NO_PATH_FOUND`.

- Added the `/v0/transit/distance` response schema and integration coverage. The new tests exercise the endpoint behavior through the app: adjacent stations, same-station distance, symmetry, two-hop routing, validation errors, unknown stations, and disconnected stations.

- Simplified distance caching to be memory-only. The previous disk-backed all-pairs cache had stale-cache risk because invalidation was not tied to the actual transit topology. The new shape keeps lazy graph loading with the `graphPromise` concurrency guard and uses a dynamic in-memory cache for requested distances only. This keeps the runtime simpler, avoids precompute work, and keeps `line_stations`/`lines` as the source of truth.

johan-t commented:
@codex review this PR again

johan-t committed:
feat: do not penalize line switches

The point of this endpoint is to return how fast theorethically an inspector could reach your station. Inspectors do not really have a prefrence for staying on one line since they get off each stop anyways, so penalizing line switch makes sense when building for consumers but not for this use case
